subprojects/gstreamer/plugins/tracers/gstframerate.c

```diff
@@ -96,13 +122,206 @@ is_target_video_sink (GstElement *element)
+
+  g_return_if_fail (socket != NULL);
+
+  if (!g_socket_close (socket, &error)) {
+    GST_ERROR_OBJECT (self, "Error closing socket: %s", error->message);
+  }
+  g_object_unref (socket);
+  if (error)
+    g_error_free (error);
+}
+
+static GIOStream *
+gst_framerate_client_socket_connect (GstFramerateTracer *self, GSocket *socket)
+{
+  GSocketAddress *addr;
+  GSocketConnectable *connectable;
+  GSocketAddressEnumerator *enumerator;
+  GSocketAddress *address;
+  GIOStream *connection;
+  GCancellable *cancellable = NULL;
+  GError *error = NULL;
+
+  g_return_val_if_fail (socket != NULL, NULL);
+
+  addr = socket_address_from_string (self->unix_socket_path);
+  if (addr == NULL) {
+    GST_ERROR_OBJECT (self, "Could not create a socket address at '%s'",
+        self->unix_socket_path);
+    return NULL;
+  }
+
+  connectable = G_SOCKET_CONNECTABLE (addr);
+  enumerator = g_socket_connectable_enumerate (connectable);
+  g_object_unref (connectable);
+
+  while (TRUE) {
+    address =
+        g_socket_address_enumerator_next (enumerator, cancellable, &error);
+    if (address == NULL) {
+      if (error == NULL)
+        GST_DEBUG_OBJECT (self, "No more addresses to try");
+      else {
+        GST_DEBUG_OBJECT (self, "%s", error->message);
+        g_error_free (error);
+      }
+      return NULL;
+    }
+
+    if (g_socket_connect (socket, address, cancellable, &error)) {
+      g_object_unref (address);
+      break;
+    }
+
+    GST_ERROR_OBJECT (self, "Connection to %s failed: %s, trying next",
+        g_unix_socket_address_get_path (G_UNIX_SOCKET_ADDRESS (address)),
+        error->message);
+    g_error_free (error);
+    error = NULL;
+
+    g_object_unref (address);
+  }
+  g_object_unref (enumerator);
+
+  GST_DEBUG_OBJECT (self, "Connected to %s",
+      g_unix_socket_address_get_path (G_UNIX_SOCKET_ADDRESS (address)));
+
+  connection =
+      G_IO_STREAM (g_socket_connection_factory_create_connection (socket));
+  return connection;
+}
+
+static void
+gst_framerate_client_socket_disconnect (GstFramerateTracer *self,
+    GIOStream *connection)
+{
+  GCancellable *cancellable = NULL;
+  GError *error = NULL;
+
+  g_return_if_fail (connection != NULL);
+
+  if (!g_io_stream_close (connection, cancellable, &error)) {
+    GST_ERROR_OBJECT (self, "Error closing connection.");
+    return;
+  }
+
+  g_object_unref (connection);
+}
+
+static void
+gst_framerate_client_socket_send_internal (GstFramerateTracer *self,
+    GIOStream *connection, gchar *data, gsize len)
+{
+  GOutputStream *ostream;
+  GError *error = NULL;
+  gsize size;
+
+  ostream = g_io_stream_get_output_stream (connection);
+  size = g_output_stream_write (ostream, data, len, NULL, &error);
+
+  if (error != NULL) {
+    GST_ERROR_OBJECT (self, "Failed to send (%ld): %s. Error: %s",
+        len, data, error->message);
+    g_error_free (error);
+  }
+  if (size != len)
+    GST_ERROR_OBJECT (self, "Failed to write full data. Wrote %ld bytes", size);
```

Comment:
shouldn't it be

```c
while (len != 0) {
  gsize size;

  size = g_output_stream_write (ostream, data, len, NULL, &error);

  if (error != NULL) {
    GST_ERROR_OBJECT (self, "Failed to send (%ld): %s. Error: %s",
        len, data, error->message);
    break;
  }
  
  g_return_if_fail (len >= size);
  len -= size;
  data += size;
}
```

Not writing all the data is a valid case:
> On success, the number of bytes written to the stream is returned. It is not an error if this is not the same as the requested size, as it can happen e.g. on a partial I/O error, or if there is not enough storage in the stream. All writes block until at least one byte is written or an error occurs; 0 is never returned (unless count is 0).